ext/fileinfo/fileinfo.c

```diff
@@ -275,181 +275,198 @@ PHP_FUNCTION(finfo_set_flags)
+	if (options) {
+		magic_setflags(magic, options);
 	}
 
-common:
+	const char *ret_val = magic_buffer(magic, ZSTR_VAL(buffer), ZSTR_LEN(buffer));
+
 	if (ret_val) {
 		RETVAL_STRING(ret_val);
 	} else {
 		php_error_docref(NULL, E_WARNING, "Failed identify data %d:%s", magic_errno(magic), magic_error(magic));
 		RETVAL_FALSE;
 	}
 
-clean:
-	if (mimetype_emu) {
-		magic_close(magic);
-	}
-
 	/* Restore options */
 	if (options) {
 		magic_setflags(magic, options);
 	}
 }
 /* }}} */
 
-/* {{{ Return information about a file. */
-PHP_FUNCTION(finfo_file)
+/* Return content-type for file */
+PHP_FUNCTION(mime_content_type)
 {
-	_php_finfo_get_type(INTERNAL_FUNCTION_PARAM_PASSTHRU, FILEINFO_MODE_FILE, 0);
-}
-/* }}} */
+	zval *path_or_stream;
+	zend_string *path = NULL;
+	php_stream *stream = NULL;
+	struct magic_set *magic = NULL;
 
-/* {{{ Return information about a string buffer. */
-PHP_FUNCTION(finfo_buffer)
-{
-	_php_finfo_get_type(INTERNAL_FUNCTION_PARAM_PASSTHRU, FILEINFO_MODE_BUFFER, 0);
-}
-/* }}} */
+	/* mime_content_type(..) emulation */
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "z", &path_or_stream) == FAILURE) {
+		RETURN_THROWS();
+	}
 
-/* {{{ Return content-type for file */
-PHP_FUNCTION(mime_content_type)
-{
-	_php_finfo_get_type(INTERNAL_FUNCTION_PARAM_PASSTHRU, -1, 1);
+	switch (Z_TYPE_P(path_or_stream)) {
+		case IS_STRING:
+			path = Z_STR_P(path_or_stream);
+			if (UNEXPECTED(ZSTR_LEN(path) == 0)) {
+				zend_argument_must_not_be_empty_error(1);
+				RETURN_THROWS();
+			}
+			if (UNEXPECTED(zend_str_has_nul_byte(path))) {
+				zend_argument_type_error(1, "must not contain any null bytes");
+				RETURN_THROWS();
+			}
+			break;
+
+		case IS_RESOURCE:
+			php_stream_from_zval(stream, path_or_stream);
+			break;
+
+		default:
+			zend_argument_type_error(1, "must be of type resource|string, %s given", zend_zval_value_name(path_or_stream));
+			RETURN_THROWS();
+	}
+
+	magic = magic_open(MAGIC_MIME_TYPE);
+	if (UNEXPECTED(magic == NULL)) {
+		php_error_docref(NULL, E_WARNING, "Failed to load magic database");
+		RETURN_FALSE;
+	}
+
+	if (UNEXPECTED(magic_load(magic, NULL) == -1)) {
+		php_error_docref(NULL, E_WARNING, "Failed identify data %d:%s", magic_errno(magic), magic_error(magic));
+		magic_close(magic);
+		RETURN_FALSE;
+	}
+
+	const char *ret_val;
+	if (path) {
+		php_stream_context *context = php_stream_context_from_zval(NULL, false);
+		ret_val = php_fileinfo_from_path(magic, path, context);
+	} else {
+		/* remember stream position for restoration */
+		zend_off_t current_stream_pos = php_stream_tell(stream);
+		php_stream_seek(stream, 0, SEEK_SET);
+
+		ret_val = magic_stream(magic, stream);
+		if (UNEXPECTED(ret_val == NULL)) {
+			php_error_docref(NULL, E_WARNING, "Failed identify data %d:%s", magic_errno(magic), magic_error(magic));
+		}
+
+		php_stream_seek(stream, current_stream_pos, SEEK_SET);
+	}
+
+	if (ret_val) {
+		RETVAL_STRING(ret_val);
+	} else {
+		RETVAL_FALSE;
+	}
+	magic_close(magic);
```

Comment:
I'd move this magic_close before the if so that they can be RETURN_ instead of RETVAL_

Reply:
I am getting a UAF if I move it before the RETVAL_STRING.

Reply:
I see, the string is bound to the lifetime of the magic instance, instead of to global data most likely. Ok then, leave it like this then.

Reply:
Might add a comment just so that this not forgotten.